### PR TITLE
ENH: Adds random sampling

### PR DESCRIPTION
**Pull request by jfkirk**
1. Adds `is_sample_based` parameter to loss graphs.
2. Adds logic to generate random sample indices per batch during fitting.
3. Adds logic to gather the predictions from the randomly sampled items for loss.
4. Adds random sample logic to WMRB losses.
5. Adds test for the random sample gathering.
6. Fixes WMRB logic.
7. Removes SeparationLossGraph.
8. Various small refactors.

### CODE REVIEW

**Comment by jfkirk** [comment]
cc @mesanders again

File: tensorrec/loss_graphs.py [hunk 1]
```diff
@@ -54,59 +63,50 @@ def loss_graph(self, tf_interactions, tf_prediction, **kwargs):
...
     """
     is_dense = True
+    is_sample_based = True
 
-    def loss_graph(self, tf_interactions, tf_prediction, **kwargs):
+    def loss_graph(self, tf_prediction, tf_interactions, tf_sample_predictions, **kwargs):
+
+        # WMRB expects bounded predictions
+        tanh_prediction = tf.nn.sigmoid(tf_prediction)
+        tanh_sample_prediction = tf.nn.sigmoid(tf_sample_predictions)
```
Comment (jcauteru) [inline tensorrec/loss_graphs.py]: Think it is ok to hard code with sig? or maybe warn is prediction value is unbounded

Comment (jcauteru) [inline tensorrec/loss_graphs.py]: naming `tanh...`

File: tensorrec/loss_graphs.py [hunk 2]
```diff
@@ -54,59 +63,50 @@ def loss_graph(self, tf_interactions, tf_prediction, **kwargs):
...
+        # [ n_positive_interactions, n_sampled_items ]
+        mapped_predictions_sample_per_interaction = tf.gather(params=tf_sample_predictions,
+                                                              indices=tf.transpose(positive_interaction_indices)[0])
 
-        sampled_margin_rank = (n_items - (n_items * positive_predictions)
-                               + mapped_predictions_sum_per_user + irrelevant_item_indicator)
+        # [ n_positive_interactions, n_sampled_items ]
+        summation_term = tf.abs(1.0
+                                - tf.expand_dims(positive_predictions, axis=1)
+                                + mapped_predictions_sample_per_interaction)
```
Comment (jcauteru) [inline tensorrec/loss_graphs.py]: What about the indicator? assuming this is rolled into `positive_predictions`, `mapped_predictions_sample_per_interaction`?

File: tensorrec/recommendation_graphs.py [hunk 3]
```diff
@@ -62,8 +62,8 @@ def alignment(tf_user_representation, tf_item_representation):
     :param tf_item_representation:
     :return:
     """
-    normalized_users = tf.nn.l2_normalize(tf_user_representation, 0)
-    normalized_items = tf.nn.l2_normalize(tf_item_representation, 0)
+    normalized_users = tf.nn.l2_normalize(tf_user_representation, 1)
+    normalized_items = tf.nn.l2_normalize(tf_item_representation, 1)
```
Comment (jcauteru) [inline tensorrec/recommendation_graphs.py]: :clap:

File: tensorrec/recommendation_graphs.py [hunk 4]
```diff
@@ -94,6 +94,24 @@ def bias_prediction_serial(tf_prediction_serial, tf_projected_user_biases, tf_pr
...
+    """
+    prediction_shape = tf.shape(tf_prediction)
+    flattened_prediction = tf.reshape(tf_prediction, shape=[prediction_shape[0] * prediction_shape[1]])
+
+    indices_shape = tf.shape(tf_sampled_item_indices)
+    flattened_indices = tf.reshape(tf_sampled_item_indices, shape=[indices_shape[0] * indices_shape[1]])
+
+    gathered_predictions = tf.gather(params=flattened_prediction, indices=flattened_indices)
+    reshaped_gathered_predictions = tf.reshape(gathered_predictions, shape=indices_shape)
+    return reshaped_gathered_predictions
```
Comment (jcauteru) [inline tensorrec/recommendation_graphs.py]: LTGM

File: tensorrec/tensorrec.py [hunk 5]
```diff
@@ -351,10 +367,12 @@ def fit(self, interactions, user_features, item_features, epochs=100, learning_r
                          alpha=alpha,
                          verbose=verbose,
                          out_sample_interactions=out_sample_interactions,
-                         user_batch_size=user_batch_size)
+                         user_batch_size=user_batch_size,
+                         n_sampled_items=n_sampled_items)
 
     def fit_partial(self, interactions, user_features, item_features, epochs=1, learning_rate=0.1,
-                    alpha=0.00001, verbose=False, out_sample_interactions=None, user_batch_size=None):
+                    alpha=0.00001, verbose=False, out_sample_interactions=None, user_batch_size=None,
```
Comment (jcauteru) [inline tensorrec/tensorrec.py]: pretty small default alpha?

File: test/test_loss_graphs.py [hunk 6]
```diff
@@ -30,26 +30,18 @@ def test_rmse_dense_loss_biased(self):
...
 
     def test_wmrb_alignment_loss(self):
         model = TensorRec(loss_graph=WMRBAlignmentLossGraph)
-        model.fit(self.interactions, self.user_features, self.item_features, epochs=5)
+        model.fit(self.interactions, self.user_features, self.item_features, epochs=5, n_sampled_items=10)
 
     def test_wmrb_alignment_loss_biased(self):
         model = TensorRec(loss_graph=WMRBAlignmentLossGraph, biased=True)
-        model.fit(self.interactions, self.user_features, self.item_features, epochs=5)
+        model.fit(self.interactions, self.user_features, self.item_features, epochs=5, n_sampled_items=10)
```
Comment (jcauteru) [inline test/test_loss_graphs.py]: maybe add another test with different n_sampled_items? Make sure that it changes at least?

**Comment by jcauteru** [comment]
Some general comments. Gave it a once over, can look again. Might be good to have @mesanders give it s second set of eyes

File: tensorrec/loss_graphs.py (hunk 1 above)
Comment (jfkirk) [reply tensorrec/loss_graphs.py]: Good catch!

Comment (jfkirk) [reply tensorrec/loss_graphs.py]: I think hard-coding is okay given that tf_prediction would be an unbounded dot product in all cases and could violate the [0, 1] that WMRB expects. It'd be tough to warn when prediction is unbounded at graph build time.

File: tensorrec/loss_graphs.py (hunk 2 above)
Comment (jfkirk) [reply tensorrec/loss_graphs.py]: If we (naiively) assume that the indicator is 1 for all sampled items, and since its multiplied in, I just didn't write it in here. I'll add a TODO noting that the irrelevant item indicator isn't in place.

File: tensorrec/tensorrec.py (hunk 5 above)
Comment (jfkirk) [reply tensorrec/tensorrec.py]: It is -- I wanted to have a small, non-zero alpha in place but because the relationship between weight normalization and the loss function is strong, I didn't want to have the alpha so high that it destroys performance for some loss functions.

File: test/test_loss_graphs.py (hunk 6 above)
Comment (jfkirk) [reply test/test_loss_graphs.py]: Good point -- I think a test of the `n_random_samples` should live in `TensorRecTestCase`, since the logic is contained in the `TensorRec` object.

At the very least, these loss graph tests should do more than just make sure the model fits.

**Comment by jfkirk** [comment]
Agreed - would love to get @mesanders 's take.

Also I've re-added SeparationLossGraph and added SeparationDenseLossGraph. SeparationDense is now the top performing loss on Movielens 10k, getting over 60% recall at 100. For... reasons. ¯\_(ツ)_/¯

File: tensorrec/tensorrec.py [hunk 7]
```diff
@@ -294,16 +298,25 @@ def _build_tf_graph(self, n_user_features, n_item_features):
...
+            )
+            tf_sample_alignments = gather_sampled_item_predictions(
+                tf_prediction=self.tf_alignment, tf_sampled_item_indices=self.tf_sampled_item_indices
+            )
+            loss_graph_kwargs.update({'tf_sample_predictions': tf_sample_predictions,
+                                      'tf_sample_alignments': tf_sample_alignments})
 
         # Build loss graph
-        self.tf_basic_loss = self.loss_graph_factory().loss_graph(**loss_graph_kwargs)
+        self.tf_basic_loss = self.loss_graph().loss_graph(**loss_graph_kwargs)
```
Comment (mesanders) [inline tensorrec/tensorrec.py]: I know it's due to the refactor and makes better sense that the reference to the class be called loss_graph instead of loss_graph_factory, but 
`self.loss_graph().loss_graph(**)` 
Just feels a bit wonky, but not worth changing.

Comment (jfkirk) [reply tensorrec/tensorrec.py]: Maybe `loss_graph().connect_loss_graph()`?
